---
title: "Analyze the ProToCol dataset"
output:
  html_document:
    df_print: paged
---

```{r}
require(knitr)
require(ggplot2)
require(reshape2)
```

# Introduction

* Next generation sequencing data, so not microarray
* Used as validation set in the publication below
https://www.researchgate.net/publication/330971406_Widespread_and_Functional_RNA_Circularization_in_Localized_Prostate_Cancer

* Used the protocol defined below
https://ega-archive.org/studies/EGAS00001002816

* 41 normal tissue samples
* 51 prostate cancer tissue samples

# Dataset exploration

```{r load the dataset}
ngs = read.delim("GeneExpressionData/Galaxy37-[edgeR_DGE_on_2__design_matrix_prostate_unpaired.txt_-_differentially_expressed_genes].tabular.annotated.txt", stringsAsFactors = F, row.names = 1)
colnames(ngs) = c("Ensembl_ID", "Gene_Symbol", "log2FC", "log2CPM", "LR", "PValue", "FDR")

ngs$FC = 2^ngs$log2FC
ngs$Ensembl_ID_no_Dot = sapply(ngs$Ensembl_ID, function(x){return(unlist(strsplit(x, "\\."))[1])})
```

* EdgeR was used to create the table below
* By default, EdgeR uses a log2 for the fold change
* Usual FDR cutoffs: 0.01, 0.001
* Column names indicate:
* Unknown what kind of log the logCPM is

* CPM stands for "Counts per million"
* LR stands for likelihood ratio?
* Pvalue is two-sided

In total, the dataset contains `r nrow(ngs)` entries.
These also include pseudogenes and non-coding sequences.
Some of these have also been found to be associated with prostate cancer.
I'm not yet sure whether to also include these, that remains to be discussed.
For now, I'll analyse all entries in the dataset.
Once we have made a decision, I can analyze the dataset again. 

The quantiles of the values of the logFC can be seen below:

`r kable(data.frame(Values = quantile(ngs$logFC)))`

The mean logFC is `r mean(ngs$logFC)`, and the median value is `r median(ngs$logFC)`.

Below, I plot a histogram of the values: 

`r hist(ngs$log2FC, breaks = 100, main = "Fold-change of the sequences in the ProToCol data", xlab = "2logFC")`

# Filter by protein coding genes

To only include protein-coding genes, I download the list of protein coding genes (so not transcripts, which was also an option) from Ensembl's BioMart for the GRCh38.p12 build (which should be the same as the build used for the ProToCol study results).
These set of protein coding genes was downloaded on 04-04-2019. 

```{r protein coding genes}
protein_coding_genes = read.csv("GeneExpressionData/ensembl biomart_export protein coding genes 04-04-2019.txt", stringsAsFactors = F)
ngs_pc = ngs[ngs$Ensembl_ID_no_Dot %in% protein_coding_genes$Gene.stable.ID, ]
```

# Setting the cutoffs
As recommended, I use the FDR to identify differentially expressed genes between the normal and the PCa tissue.
I test multiple cutoffs, to examine how many genes remain with every FDR value. 

```{r set cutoffs}
cutoffs = 0.1^(0:25)
foldchanges = seq(0, 1, 0.05)

counts = expand.grid(cutoffs, foldchanges)
colnames(counts) = c("cutoff", "fold.change")

for(i in 1:nrow(counts)){
  counts[i, "All"] = nrow(ngs[ngs$FDR < counts$cutoff[i] & (ngs$FC > 1 + counts$fold.change[i] | ngs$FC < 1 - counts$fold.change[i]), ])
  counts[i, "Protein.coding"] = nrow(ngs_pc[ngs_pc$FDR < counts$cutoff[i] & (ngs_pc$FC > 1 + counts$fold.change[i] | ngs_pc$FC < 1 - counts$fold.change[i]), ])
}

wireframe(Protein.coding ~ fold.change*log10(cutoff), data = rev(counts), screen = list(z = -60, x = -60), drape = T, par.settings = list(axis.line = list(col = "transparent"), box.3d = list(col=c(1,NA,NA,1,1,NA,1,1,1))), scales=list(arrows=F,col=1), ylab = "False Discovery\n Rate cutoff", xlab = "Fold change\n cutoff", zlab = "Count") 
```

```{r plot counts, echo = F}
counts2 = data.frame(cutoff = c(cutoffs, cutoffs), Type = c(rep("All", length(cutoffs)), rep("Protein coding", length(cutoffs))), Value = c(counts$All, counts$Protein.coding))
ggplot(data = counts2) + geom_bar(aes(x = cutoff, y = Value, fill = Type), stat = "identity", position = "dodge") + scale_x_continuous(trans = "log10") + ylim(0,2500)
```

```{r create surface plot}
require(lattice)


```


# Load reference sets

```{r refsets}
small = read.csv("Reference sets/RefSet Farashi and Hazelett Small.csv", stringsAsFactors = F, header = F)
farashi = read.delim("Reference sets/Farashi SM1 genes.csv", stringsAsFactors = F, header = F)
farashi = unlist(lapply(farashi$V1, function(x){strsplit(x, ";")})) # Split up the multiple genes per line
farashi = unlist(lapply(farashi, function(x){strsplit(x, ", ")})) # Split up the multiple genes per line
farashi = gsub(" ", "", farashi) # Remove any remaining white spaces
farashi = unique(farashi)
```

Please note that Excel keeps making SEPT2 into a date. This should be fixed manually.
The (PSA) behind KLK3 was removed, as well as the cross annotations behind MSMB1, 2, and NCOA4-1, 3.
The "FoxA1 binding" part behind c-MYC was removed aswell, as was the "(lncRNA)" part behind PCAT1.


Both "MSMB1" and "MSMB2" are not available in the ProToCol gene list.
According to EntrezGene and the HGNC, only "MSMB" exists. 
I'll therefore assume these entries are variants this gene, and abstract both of them to this gene.

```{r}
farashi[farashi %in% c("MSMB1", "MSMB2")] = "MSMB"
```

NCOA4-1 and NCOA4-3 are also not in the ProToCol list.
I was unable to find a human gene for both entries in EntrezGene, and the closest gene in the HGNC data are "NCOA4P1" and "NCOA4P3".

```{r}
farashi[farashi == "NCOA4-1"] = "NCOA4P1"
farashi[farashi == "NCOA4-3"] = "NCOA4P3"
```

Neither of these are in the NGS data as well.

According to HGNC and EntrezGene, "ANKRD5" is now "ANKEF1", so I'll replace that value.

```{r}
farashi[farashi == "ANKRD5"] = "ANKEF1"
```


When I checked "C6orf228" in entrezGene, I was forwarded to "SMIM13", which is in the list.
Based on the EntrezGene information, I therefore replace C6orf228 with "SMIM13"

```{r}
farashi[farashi == "C6orf228"] = "SMIM13"
```

According to HGNC, "c-MYC" is a synonym of "MYC", so I'll replace that value.

```{r}
farashi[farashi == "c-MYC"] = "MYC"
```

"HoxB13" is should probably all be capitals.
When all letters are capitalized, it is available in both the HGNC data as well as the ProToCol data. 

```{r}
farashi[farashi == "HoxB13"] = "HOXB13"
```

According to both EntrezGene and HGNC, "LASS2" has been renamed to "CERS2".
"CERS2" is included in the NGS results.

```{r}
farashi[farashi == "LASS2"] = "CERS2"
```

According to EntrezGene and HGNC, "C10orf32" has been renamed "BORCS7". 
"BORCS7" is also included in the NGS results.

```{r}
farashi[farashi == "C10orf32"] = "BORCS7"
```

According to EntrezGene, "LOC100505761" is "RPARP-AS1".
I was unable to very this in HGNC. 
"RPARP-AS1" is available in the NGS results.

```{r}
farashi[farashi == "LOC100505761"] = "RPARP-AS1"
```

According to both EntrezGene and HGNC, "LOC100505495" has been mapped to "PCAT19", which is in the NGS results.

```{r}
farashi[farashi == "LOC100505495"] = "PCAT19"
```

According to both EntrezGene and HGNC, "WDR52" has been mapped to "CFAP44", which is in the NGS results.

```{r}
farashi[farashi == "WDR52"] = "CFAP44"
```

Both "LOC90246", "LOC729603", and "LOC284581" were available in EntrezGene, but not in the HGNC data.
Based on EntrezGene, I would say these identifiers are correct and not outdated.

I was unable to find any information about "NCR00171" in EntrezGene and HGNC, or even on google.

According to both EntrezGene and the HGNC, "HCG4P6" has been mapped to "HCG4B", which is in the NGS results.

```{r}
farashi[farashi == "HCG4P6"] = "HCG4B"
```

According to EntrezGene, "LOC285830" is "HLA-F-AS1".
I was unable to very this in HGNC. 
"HLA-F-AS1" is available in the NGS results.

```{r}
farashi[farashi = "LOC285830"] = "HLA-F-AS1"
```

According to both EntrezGene and HGNC, "RAB7L1" has been mapped to "RAB29", which is in the NGS results.

```{r}
farashi[farashi == "RAB7L1"] = "RAB29"
```

According to EntrezGene, "LOC284578" is "MFSD4A-AS1".
I was unable to very this in HGNC. 
"MFSD4A-AS1" is not available in the NGS results.

```{r}
farashi[farashi = "LOC284578"] = "MFSD4A-AS1"
"MFSD4A-AS1" %in% ngs$Gene_Symbol
```

According to both EntrezGene and HGNC, "AGAP7" has been mapped to "AGAP7P", which is in the NGS results.

```{r}
farashi[farashi == "AGAP7"] = "AGAP7P"
```

According to both EntrezGene and HGNC, "C2orf43" has been mapped to "LDAH", which is in the NGS results.

```{r}
farashi[farashi == "C2orf43"] = "LDAH"
```

# Constructing the reference set

It appears that not all genes that are mentioned in Farashi's paper are also mentioned in the supplemental materials.
I therefore combine both lists, to create an ensemble list which we can use as a reference set during further analyses.

```{r}
refset = unique(c(small$V1, farashi)) 
```

In total, the reference set contains `r length(refset)` genes, of which `r length(intersect(refset, ngs$Gene_Symbol))` are in the NGS results, and `r length(intersect(refset, ngs_pc$Gene_Symbol))`are in the protein coding part of the NGS results.

I now visualize the effect of using different FDR and FC cutoffs on the NGS data on the reference set.

```{r }
for(i in 1:nrow(counts)){
  counts[i, "All"] = length(intersect(ngs$Gene_Symbol[ngs$FDR < counts$cutoff[i] & (ngs$FC > 1 + counts$fold.change[i] | ngs$FC < 1 - counts$fold.change[i])], refset))
  counts[i, "Protein.coding"] = length(intersect(ngs_pc$Gene_Symbol[ngs_pc$FDR < counts$cutoff[i] & (ngs_pc$FC > 1 + counts$fold.change[i] | ngs_pc$FC < 1 - counts$fold.change[i])], refset))
}

wireframe(Protein.coding ~ fold.change*log10(cutoff), data = rev(counts), screen = list(z = -60, x = -60), drape = T, par.settings = list(axis.line = list(col = "transparent"), box.3d = list(col=c(1,NA,NA,1,1,NA,1,1,1))), scales=list(arrows=F,col=1), ylab = "False Discovery\n Rate cutoff", xlab = "Fold change\n cutoff", zlab = "Count") 
```

